File: src/main/java/de/nekeras/borderless/DesktopEnvironment.java
package de.nekeras.borderless;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.lwjgl.system.Platform;

import de.nekeras.borderless.fullscreen.NativeFullscreen;

/**
 * The desktop environment we are running on, retrievable by {@link #get()}.
 */
public enum DesktopEnvironment {

    /**
     * The Windows desktop environment.
     */
    WINDOWS,

    /**
     * X11 window system used by Linux distributions.
     */
    X11,

    /**
     * An unknown desktop environment, this should always use {@link NativeFullscreen}.
     */
    GENERIC;

    private static final String LINUX_WINDOW_SYSTEM_VARIABLE = "XDG_SESSION_TYPE";
    private static final DesktopEnvironment CURRENT;
    private static final Logger log = LogManager.getLogger();

    static {
        log.info("Determining desktop environment");

        DesktopEnvironment current = null;

        switch (Platform.get()) {
            case WINDOWS:
                current = WINDOWS;
                break;
            case LINUX:
                String result = System.getenv(LINUX_WINDOW_SYSTEM_VARIABLE);

                if ("x11".equalsIgnoreCase(result)) {
                    current = X11;
                } else {
                    // Also treats Wayland as unknown as this seems buggy with LWJGL
                    current = GENERIC;
                    log.warn("Unknown window system: {}", result);
                }

                break;
            case MACOSX:
                current = GENERIC;
                break;
        }

        CURRENT = current;
        log.info("Found desktop environment {}", CURRENT);
    }

    /**
     * The current desktop environment Minecraft is running on.
     *
     * @return The desktop environment
     */
    public static DesktopEnvironment get() {
        return CURRENT;
    }

}

File: src/main/kotlin/de/nekeras/borderless/DesktopEnvironment.kt
package de.nekeras.borderless

import de.nekeras.borderless.fullscreen.BorderlessFullscreen
import de.nekeras.borderless.fullscreen.FullscreenMode
import de.nekeras.borderless.fullscreen.NativeFullscreen
import de.nekeras.borderless.fullscreen.NativeNonIconifyFullscreen
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.lwjgl.system.Platform

/**
 * Possible desktop environments Minecraft can be run on. Each environment may have a [bestFullscreenMode]
 * fullscreen mode that should be used when executed in such an environment.
 */
enum class DesktopEnvironment(val bestFullscreenMode: FullscreenMode) {

    /**
     * The Windows operating system desktop environment.
     */
    WINDOWS(BorderlessFullscreen),

    /**
     * X11 window system used by Linux operating systems.
     */
    X11(NativeNonIconifyFullscreen),

    /**
     * Wayland display server used by Linux operating system.
     */
    WAYLAND(NativeFullscreen),

    /**
     * A generic desktop environment that does not fit in any other category.
     */
    GENERIC(NativeFullscreen);

    companion object {

        private const val LINUX_WINDOW_SYSTEM_VARIABLE = "XDG_SESSION_TYPE";
        private const val X11_NAME = "x11"
        private const val WAYLAND_NAME = "wayland"

        private val log: Logger = LogManager.getLogger(DesktopEnvironment::class.java)

        @JvmStatic
        val current: DesktopEnvironment =
            when (Platform.get()) {
                Platform.WINDOWS -> WINDOWS
                Platform.LINUX -> when (System.getenv(LINUX_WINDOW_SYSTEM_VARIABLE)) {
                    X11_NAME -> X11
                    WAYLAND_NAME -> WAYLAND
                    else -> GENERIC
                }
                else -> GENERIC // Treat macOS as Generic
            }.also {
                log.info("Found desktop environment $it")
            }

    }

}
